push_config.go

```diff
@@ -0,0 +1,176 @@
+	}
+}
+
+// NewPushConfig returns a new PushConfig.
+func NewPushConfig() *PushConfig {
+
+	return &PushConfig{
+		Identities:            map[string][]EventType{},
+		IdentityFilters:       map[string]string{},
+		parsedIdentityFilters: map[string]*Filter{},
+	}
+}
+
+// SetParameter sets the values of the parameter with the given key.
+func (f *PushConfig) SetParameter(key string, values ...string) {
+
+	if f.Params == nil {
+		f.Params = url.Values{}
+	}
+
+	f.Params[key] = values
+}
+
+// Parameters returns a copy of all the parameters.
+func (f *PushConfig) Parameters() url.Values {
+
+	if f.Params == nil {
+		return nil
+	}
+
+	out := url.Values{}
+	for k, v := range f.Params {
+		out[k] = v
+	}
+
+	return out
+}
+
+// FilterIdentity adds the given identity for the given eventTypes in the PushConfig.
+func (f *PushConfig) FilterIdentity(identityName string, eventTypes ...EventType) {
+
+	f.Identities[identityName] = eventTypes
+}
+
+// ParseIdentityFilters does something...
+//
+// TODO:
+//  - add a proper comment explaining what this API does
+//  - add unit tests
+func (f *PushConfig) ParseIdentityFilters() error {
+
+	for identity, unparsedFilter := range f.IdentityFilters {
+		if _, found := f.Identities[identity]; !found {
+			return fmt.Errorf("elemental: cannot declare an identity filter on %q as that was not declared in 'Identities'", identity)
+		}
+
+		filter, err := NewFilterParser(unparsedFilter).Parse()
+		if err != nil {
+			return fmt.Errorf("elemental: unable to parse filter %q: %s", unparsedFilter, err)
+		}
+
+		f.parsedIdentityFilters[identity] = filter
+	}
+
+	return nil
+}
+
+// IsFilteredOut returns true if the given Identity is not part of the PushConfig's Identity mapping
+func (f *PushConfig) IsFilteredOut(identityName string, eventType EventType) bool {
+
+	// if the identities list is empty, we filter nothing.
+	if len(f.Identities) == 0 {
+		return false
+	}
+
+	// If it contains something, but not the identity, we filter out.
+	types, ok := f.Identities[identityName]
+	if !ok {
+		return true
+	}
+
+	// If there is no event types defined we don't filter
+	if len(types) == 0 {
+		return false
+	}
+
+	// If if there are some event types defined, we don't filter out
+	// if the current event type is in the list.
+	for _, t := range types {
+		if t == eventType {
+			return false
+		}
+	}
+
+	return true
+}
+
+// Duplicate duplicates the PushConfig.
+func (f *PushConfig) Duplicate() *PushConfig {
+
+	pc := NewPushConfig()
+
+	for id, types := range f.Identities {
+		pc.FilterIdentity(id, types...)
+	}
+
+	for id, f := range f.IdentityFilters {
+		pc.IdentityFilters[id] = f
+	}
+
+	for id, f := range f.parsedIdentityFilters {
+		pc.parsedIdentityFilters[id] = f
+	}
+
+	for k, v := range f.Params {
+		pc.SetParameter(k, v...)
+	}
+
+	return pc
+}
+
+func (f *PushConfig) String() string {
+
+	return fmt.Sprintf("<pushconfig identities:%s identityfilters:%s>", f.Identities, f.IdentityFilters)
```

Comment:
**Note this is a breaking (behavioural) change**

Previously:

	return fmt.Sprintf("<pushfilter identities:%s>", f.Identities)


Now: 

	return fmt.Sprintf("<pushconfig identities:%s identityfilters:%s>", f.Identities, f.IdentityFilters)